### PR TITLE
make activity and rules button hideable

### PR DESCRIPTION
**Pull request by xithan**
- activity and rules button can be hidden via read permissions for the cards `*activity toolbar button` and `*rules toolbar button`
- add new role: "Anyone with Role" (needed by ARB deck)

### CODE REVIEW

File: card/lib/card/model/save_helper.rb [hunk 1]
```diff
@@ -10,11 +10,9 @@ module Model
...
     module SaveHelper
       def as_user user_name
-        current = Card::Auth.current_id
-        Card::Auth.current_id = Card.fetch_id user_name
-        yield
-      ensure
-        Card::Auth.current_id = current
+        Card::Auth.with current_id: Card.fetch_id(user_name) do
+          yield
+        end
```
Comment (ethn) [inline card/lib/card/model/save_helper.rb]: I know this is old, but this isn't usually the way we use "as", right?  Doesn't "as" change the permissions but not the current user?  (Eg, if I'm PK and I do something "as_bot", I get bot permissions but the action is still performed by PK.

File: card/mod/account/set/right/account.rb [hunk 2]
```diff
@@ -57,12 +57,27 @@ def token_path_opts
...
+    #   ]
+    # )
     %({{+#{Card[:email].name}|titled;title:email}}
-     {{+#{Card[:password].name}|titled;title:password}})
+      {{+#{Card[:password].name}|titled;title:password}})
   end
 
   view :edit do
     voo.structure = true
+    voo.edit_structure = [[:email, "email"], [:password, "password"]]
```
Comment (ethn) [inline card/mod/account/set/right/account.rb]: did the previous implementation stop working, or is this just more idiomatic?

File: card/db/migrate_core_cards/20170515101260_add_advanced_edit_cards.rb [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# -*- encoding : utf-8 -*-
+
+class AddAdvancedEditCards < Card::Migration::Core
+  def up
+    ensure_card "*activity toolbar button", codename: "activity_toolbar_button"
+    ensure_card "*rules toolbar button", codename: "rules_toolbar_button"
+    ensure_card "Anyone With Role",
+                codename: "anyone_with_role",
+                type_id: Card::RoleID
```
Comment (ethn) [inline card/db/migrate_core_cards/20170515101260_add_advanced_edit_cards.rb]: AnyoneWithRole is a role?  Can you explain the card representation?

File: card/mod/account/set/all/account.rb [hunk 4]
```diff
@@ -48,10 +49,13 @@ def all_roles
 end
 
 def fetch_roles
+  [Card::AnyoneSignedInID] + role_ids_from_roles_trait
+end
```
Comment (ethn) [inline card/mod/account/set/all/account.rb]: nice.

File: card/mod/core/set/all/collection.rb [hunk 5]
```diff
@@ -355,22 +355,29 @@ def construct_static_tabs tab_type
...
     columns = []
-    Card.fetch(referer).format.each_nested_field do |chunk|
+    csv_structure_card.format.each_nested_field do |chunk|
       columns << chunk.referee_name.tag
     end
     columns
   end
 
+  def csv_structure_card
+    card.rule_card(:csv_structure) || Card.fetch(row_card_names.first)
```
Comment (ethn) [inline card/mod/core/set/all/collection.rb]: new rule, right?  But no csv_structure in migration?

File: card/mod/standard/set/abstract/toolbar_split_button.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
+format :html do
+  def toolbar_split_button name, button_link_opts
+    status = active_toolbar_button == name ? "active" : ""
+    html_class = "visible-md visible-lg pull-right"
+    icon = button_link_opts.delete(:icon)
+    name_content = "&nbsp;#{name}"
+    name = icon ? glyphicon(icon) : ""
+    name += content_tag(:span, name_content.html_safe, class: html_class)
```
Comment (ethn) [inline card/mod/standard/set/abstract/toolbar_split_button.rb]: maybe name stuff needs own method?

File: card/mod/standard/set/all/rich_html/toolbar.rb [hunk 7]
```diff
@@ -188,15 +146,12 @@ def account_split_button
...
-    name = icon ? glyphicon(icon) : ""
-    name += content_tag(:span, name_content.html_safe, class: html_class)
-    button_link = button_link name, button_link_opts.merge(class: status)
-    split_button(button_link, active_toolbar_item) { yield }
+  def toolbar_button_card name
+    button_codename = "#{name}_toolbar_button".to_sym
+    return "" unless button_card = Card[button_codename]
+    with_nest_mode :normal do
+      nest button_card, view: :core
+    end
```
Comment (ethn) [inline card/mod/standard/set/all/rich_html/toolbar.rb]: I like this!

File: card/mod/standard/set/type/cardtype.rb [hunk 8]
```diff
@@ -62,8 +67,7 @@ def default_follow_set_card
 end
 
 def cards_of_type_exist?
-  # FIXME: faster test than counting all of type?
-  !new_card? && Auth.as_bot { Card.count_by_wql type_id: id } > 0
+  !new_card? && Card.where(trash: false, type_id: id).exists?
```
Comment (ethn) [inline card/mod/standard/set/type/cardtype.rb]: 👍

**Comment by xithan** [comment]
The fixes to get semaphore to pass are in the other PRs

File: card/lib/card/model/save_helper.rb (hunk 1 above)
Comment (xithan) [inline card/lib/card/model/save_helper.rb]: right, so how should I name this? `with_user` ?

File: card/mod/account/set/right/account.rb (hunk 2 above)
Comment (xithan) [inline card/mod/account/set/right/account.rb]: The previous implementation didn't work in edit mode with +*account nested in a structure.

File: card/mod/core/set/all/collection.rb (hunk 5 above)
Comment (xithan) [inline card/mod/core/set/all/collection.rb]: good point

File: card/db/migrate_core_cards/20170515101260_add_advanced_edit_cards.rb (hunk 3 above)
Comment (xithan) [inline card/db/migrate_core_cards/20170515101260_add_advanced_edit_cards.rb]: I'm not sure if I understand what you don't understand. 
If you set a permission rule to that role then anyone with a role gets access ("Anyone Signed In" doesn't count).

Comment (xithan) [inline card/db/migrate_core_cards/20170515101260_add_advanced_edit_cards.rb]: Maybe what is missing is that you should be able to explicitly assign that role to a user. You only get it automatically (like "Anyone Signed In")

Comment (xithan) [inline card/db/migrate_core_cards/20170515101260_add_advanced_edit_cards.rb]: (that feature is an ARB request)

**Pull request merged**